rfcs/45-stageless.md

```diff
@@ -0,0 +1,727 @@
+fn construction_timer_finished(timer: Res<ConstructionTimer>) -> bool {
+   timer.finished()
+}
+
+// Timers need to be ticked
+fn tick_construction_timer(mut timer: ResMut<ConstructionTimer>){
+   timer.tick();
+}
+
+// You can use queries, event readers and resources with arbitrarily complex internal logic
+fn too_many_enemies(population_query: Query<(), With<Enemy>>, population_limit: Res<PopulationLimit>) -> bool {
+   population_query.iter().count() > population_limit
+}
+
+fn main(){
+   App::new()
+   .add_plugins(DefaultPlugins)
+   .add_system(tick_construction_timer)
+   .add_system(
+      update_construction_progress
+        .after(tick_construction_timer)
+         // We can add systems with read-only system parameters as run criteria
+        .run_if(construction_timer_finished)
+   )
+   .add_system(system_meltdown_mitigation.run_if(too_many_enemies))
+   // We can use closures for simple one-off run criteria, 
+   // which automatically fetch the appropriate data from the `World`
+   .add_system(spawn_more_enemies.run_if(|difficulty: Res<Difficulty>| difficulty >= 9000))
+   // The `resource_exists` and `resource_equals` functions generate a closure that you can stick in a run criteria
+   .add_system(gravity.run_if(resource_exists(Gravity)))
+   // Run criteria can be attached to system sets, allowing them to control all systems in that set
+   .configure_set(GameSet::Physics.run_if(resource_equals(Paused(false))))
+   .run();
+}
+```
+
+There are a few important subtleties to bear in mind when working with run criteria:
+
+- when multiple run criteria are attached to the same system, the system will run if and only if all of those run criteria return true
+- each run criteria are evaluated once during each pass of the schedule
+  - this occurs just before the first system controlled by that criteria would be run
+  - the value is then shared for all other systems with that run criteria, ensuring all systems in the set stay synchronized
+
+This is important to ensure that the state of the world always matches the state expected by the system at the time it is executed.
+
+#### States
+
+**States** allow you to toggle systems on-and-off based on the value of a given resource and smoothly handle transitions between states by running cleanup and initialization systems.
+Typically, these are used for relatively complex, far-reaching facets of the game like pausing, loading assets or entering into menus.
+
+The current value of each state is stored in a resource, which must derive the `State` trait.
+These are typically (but not necessarily) enums, where each distinct state is represented as an enum variant.
+
+Each state is associated with three groups of systems:
+
+1. **While-active:** these systems run each schedule iteration if and only if the value of the state resource matches the provided value.
+   1. `app.add_system(apply_damage.run_if(state_equals(GameState::Playing))`

```

Comment:
Alrighty my new take is "why not both"?. `State::on_update` could be a SystemSet (ideally with a default ordering that ensures its commands are applied before the PostUpdate and that it runs at the appropriate time relative to on_enter/on_exit). This is where "normal state app update logic" goes. By convention this set would live before PostUpdate, but users that want to change that are welcome to.

For the more "non standard" cases (ex: after PostUpdate), the resource value checks will still work. This way the "standard convention" case has predictable (and useful) ordering defaults, allows other systems to order themselves relative to it, and can have UX consistent with `on_enter` and `on_exit`. This will almost certainly cover 99% of use cases. And the exceptional use cases can either manually define the State check in per-system RunCriteria, or build new SystemSets that follow their own arbitrary conventions / constraints.